timlee/networking-with-a-raspberry/README.md

```diff
@@ -0,0 +1,280 @@
+- `sudo nano /etc/hosts` and add a line at the bottom `192.168.100.100 badger.facts`, where the exact IP address should be replaced by that of your Raspberry Pi. The hosts file on any machine is used to convert domain searches to IP addresses, like a local DNS - in fact, it will be used here by the DNS server to resolve IP queries. We are saying that any queries for `badger.facts` should be answered with the IP address for your Pi, to reach the web server we set up earlier
+- `sudo service dnsmasq restart` will restart the server to bring these changes into effect
+
+You now need to tell your own machine that it should use this new DNS server. Go to your DNS server settings and add your Pi IP as the first option in the list, so that it is asked first (public DNS servers won't know where to find `badger.facts`, of course).
+
+Now go to your browser, type `http://badger.facts` and you should be greeted with a lovely page about badgers! Your machine is first going to your new DNS server on your Raspberry Pi to get the IP, querying that IP address and reaching the web server we set up in step 1 and receiving your lovely web page. Except... it's on http, not https. Yuck!
+
+## Step 3: SSL certificates
+
+The internet is moving away from the world of old, insecure HTTP sites towards HTTPS, an initiative driven in no small part by Google. But what's so special about an "S"?
+
+### What is HTTPS?
+
+HTTPS is HTTP over TLS (Transport Layer Security), meaning that it is fully encrypted end-to-end. This encryption is achieved with SSL (Secure Sockets Layer) certificates. Each certificate contains a public key that enables a secure handshake to take place to establish an encrypted connection before any real communication between a client (your browser) and server takes place.
+
+#### The TLS handshake
+
+The handshake to establish TLS goes (sort of) as follows:
+
+- Client: hi, server! Let's talk
+- Server: ok, but people might be listening - let's talk in code
+- Client: ooo tricky, how can we agree on a code without someone hearing what we decide?
+- Server: I've got an idea - here's a box with an open lock on it, you put the code in there, close the lock and send it back to me. Then I can use my key to open it up when it arrives.
+- Client: great idea! Here's the code I created inside the locked box. Let's use the code from now on.
+- Server: wlkejf23iojfewooioiwaejf?
+- Client: koijewf2903u0932i9evirj!!!
+
+![TCP handshake](./ssl.png)
+
+The box with an open lock on it in this over-stretched personification is a public key - part of an asymmetric key set. Data can be encrypted by anyone who has the public key (closing the lock) but it can only be decrypted by the person with the private key (opening the lock). In this case, that's the server.
+
+The 'code' generated by the client is a symmetric key - one that can be used to both encrypt and decrypt messages. Once both the client and server have the symmetric key, two-way encryption is achieved.
```

Comment:
Maybe a link to Wikipedia article about Symmetric key https://en.wikipedia.org/wiki/Symmetric-key_algorithm